kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java

```diff
@@ -515,33 +534,36 @@ private void maybeStartMergeFlushTask() {
 
   @Override
   public void stop() {
+    maybeStopExecutor(loadExecutor, "load executor");
+    maybeStopExecutor(executor, "table write executor");
+    if (upsertDelete) {
+      mergeBatches.intermediateTables().forEach(table -> {
+        logger.debug("Deleting {}", intTable(table));
+        getBigQuery().delete(table);
+      });
+    }
+
+    logger.trace("task.stop()");
+  }
+
+  private void maybeStopExecutor(ExecutorService executor, String executorName) {
+    if (executor == null) {
+      return;
+    }
+
     try {
       if (upsertDelete) {
-        mergeBatches.intermediateTables().forEach(table -> {
-          logger.debug("Deleting intermediate table {}", table);
-          getBigQuery().delete(table);
-        });
-      }
-    } finally {
-      try {
+        logger.trace("Forcibly shutting down {}", executorName);
+        executor.shutdownNow();
```

Comment:
Yeah, good question. So I noticed during manual and integration testing that the logs for basically every run included [this error message](https://github.com/apache/kafka/blob/3661f981fff2653aaf1d5ee0b6dde3410b5498db/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java#L867).

This was happening because the default task graceful shutdown timeout is [5 seconds](https://github.com/apache/kafka/blob/3661f981fff2653aaf1d5ee0b6dde3410b5498db/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java#L134), but this connector's `SinkTask`s give themselves [30 seconds](https://github.com/wepay/kafka-connect-bigquery/blob/0483de1d8b7b6e3e053777b665dc1883cdc3fda2/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L90) during `SinkTask::stop` to wait for their executors to shut down.

(I know you're familiar with Java's executors API so forgive me if you already know this next part.) The call to [ExecutorService::shutdown](https://github.com/wepay/kafka-connect-bigquery/blob/0483de1d8b7b6e3e053777b665dc1883cdc3fda2/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L371) prevents the executor from accepting any future tasks, but doesn't do anything to already-scheduled tasks (even ones that have been scheduled but are not running yet). The call to [ExecutorService::awaitTermination](https://github.com/wepay/kafka-connect-bigquery/blob/0483de1d8b7b6e3e053777b665dc1883cdc3fda2/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L372) then blocks until the timeout elapses or all of the already-scheduled tasks on that executor have completed.

If the timeout elapses (which it did frequently for me), then those tasks still end up being executed eventually--which is definitely not optimal.

I considered two possible approaches here to address this:

- Retain the existing sequence of calls to `ExecutorService::shutdown` and then `ExecutorService::awaitTermination`, and then issue a final call to `ExecutorService::shutdownNow` if the call to `awaitTermination` times out (i.e., returns `false`)
- Just invoke `ExecutorService::shutdownNow` to forcibly interrupt all running tasks immediately

I decided to go with the second approach because there doesn't seem to be any single good value we could use for the timeout with the call to `ExecutorService::awaitTermination`. The graceful task shutdown period is configurable on the worker so just hardcoding to 5 seconds still doesn't prevent zombie executor tasks from being run. We could introduce a connector config property for the executor shutdown timeout but that just doesn't seem worth it; the worst-case-scenario for forcibly terminating the executor tasks is that some data isn't available in BigQuery after the existing task is shut down and before a new round of tasks is started again, which seems preferable to allowing for possibly-duplicated data.

This new shutdown behavior is only implemented for upsert/delete mode; the existing behavior is retained otherwise.